Add tag muted-text BaseHTMLEngineProvider to render html errors

aldo-expensify opened the pull request:
### Details

The new error text:

>You have already set up this bank account on Expensify. If you'd like to add it again, please delete it [here](https://www.expensify.com/settings?param=%7B%22section%22:%22payments%22%7D).

Comes from `Web-Secure` backend and contains a link, so I'll be sending it using HTML. See related [Web-Secure PR](https://github.com/Expensify/Web-Secure/pull/2055)

Added new tag `muted-text` in BaseHTMLEngineProvider, this tag is only for setting the style `mutedTextLabel` when we find the tag. The idea is that if we receive an html error from the back, we wrap it in `<muted-text>` and render it using `RenderHTML`.

Also, we currently render all `<a>` with external urls in `href` as `AnchorForCommentsOnly`. I introduced a custom tag `<comment>` to wrap comment's html, and use it to detect if we are rendering a `<a>` within a comment or not. If we are rendering a `<a>` outside of a comment, like the error coming from the back end, we render it using a `<Text>` because `AnchorForCommentsOnly`:
- Adds a context menu that is not present in links in other parts of the app
- Doesn't align well with the rest of the text in Android

<!-- Explanation of the change or anything fishy that is going on -->

### Fixed Issues

$ https://github.com/Expensify/App/issues/5626
https://github.com/Expensify/Expensify/issues/177017

### Tests / QA

Dev: make sure you have the changes from [this PR](https://github.com/Expensify/Web-Secure/pull/2055) in you workspace

1. Create a new account in OldDot, verify the email
2. Still in OldDot, add an open VBA using data from this [SO (3)](https://stackoverflow.com/c/expensify/a/525/419). Add VBA can be found in: `Settings > Account > Payments > Add Verified Bank Account`
3. Log in with the same account in NewDot
4. Create a workspace and start the add VBA process
5. Choose to add manually the bank account, input `Routing Number: 011401533` and `Account number: 1111222233331111`
6. Click "Save & continue"
7. You should see the following error above the "Save & Continue" button:
![Screen Shot 2021-09-27 at 5 25 33 PM](https://user-images.githubusercontent.com/87341702/135002779-93294b72-cd48-4e38-99ed-287d9a5cade3.png)


### QA Steps
<!--- 
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image 
--->

### Tested On

- [x] Web
- [ ] Mobile Web
- [ ] Desktop
- [ ] iOS
- [ ] Android

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web
<!-- Insert screenshots of your changes on the web platform-->

#### Mobile Web
<!-- Insert screenshots of your changes on the web platform (from a mobile browser)-->

#### Desktop
<!-- Insert screenshots of your changes on the desktop platform-->

#### iOS
<!-- Insert screenshots of your changes on the iOS platform-->

#### Android
<!-- Insert screenshots of your changes on the Android platform-->

aldo-expensify committed:
Add tag muted-text BaseHTMLEngineProvider

When we get have an html error to display, we can wrap the error
in <muted-text> to make it have our 'mutedTextLabel' styling.

flodnv commented on src/libs/actions/BankAccounts.js:
```diff
@@ -751,7 +753,8 @@ function setupWithdrawalAccount(data) {
             } else {
                 if (response.jsonCode === 666 || response.jsonCode === 404) {
-                    error = response.message;
+                    error = response.message || response.htmlMessage;
```
I am not sure I understand this change? In general we prefer the HTML message first (on OldDot)

aldo-expensify committed:
Prioritize backend error message

Jag96 commented on src/libs/actions/BankAccounts.js:
```diff
@@ -537,9 +537,10 @@ function setFreePlanVerifiedBankAccountID(bankAccountID) {
-function showBankAccountErrorModal(errorModalMessage = null) {
-    Onyx.merge(ONYXKEYS.REIMBURSEMENT_ACCOUNT, {isErrorModalVisible: true, errorModalMessage});
+function showBankAccountErrorModal(errorModalMessage = null, isErrorModalMessageHtml = false) {
+    Onyx.merge(ONYXKEYS.REIMBURSEMENT_ACCOUNT, {isErrorModalVisible: true, errorModalMessage, isErrorModalMessageHtml});
```
Just to confirm here, if we call `merge` the value of `isErrorModalMessageHtml` will be overwritten right?

aldo-expensify replied:
I understand that is the case

Jag96 commented on src/libs/actions/BankAccounts.js:
```diff
@@ -751,7 +753,8 @@ function setupWithdrawalAccount(data) {
             } else {
                 if (response.jsonCode === 666 || response.jsonCode === 404) {
-                    error = response.message;
+                    error = response.htmlMessage || response.message;
```
Can you add a comment here explaining why we're setting htmlMessage first? Not everybody has access to the web codebase so it'd be helpful to let people know these specific cases have an HTML message and the 402/others don't.

aldo-expensify replied:
Something like:
```
// Sometimes, the response will have an error message in html format. We give priority to the html error because it has richer content.
```

About the 402/others below, I guess they could have `htmlMessage`, but we are handling those in a completely different way. We don't show their messages directly. This is inconsistent right now and out of the scope.

Jag96 replied:
I'd say something like

```
// Since these specific responses can have an error message in html format with richer content, give priority to the html error.
```

Agreed it's fine to leave the 402 alone, I just meant to add a comment like the one you proposed to highlight the difference. Good stuff 👍

aldo-expensify replied:
I used your comment, thanks 😬

aldo-expensify committed:
Add comment for htmlMessage priority over message

joelbettner commented on src/components/HTMLEngineProvider/BaseHTMLEngineProvider.js:
```diff
@@ -231,6 +235,7 @@ const renderers = {
     code: CodeRenderer,
     img: ImgRenderer,
     edited: withLocalize(EditedRenderer),
+    'muted-text': TNodeChildrenRenderer,
```
Just so I'm clear on why we are doing this, please correct me if I'm wrong...

From reading about it, I think a `TNodeChildrenRenderer` element is designed so it will render ONLY that element and its children when some data within the `TNodeChildrenRenderer` element changes. This prevents us from re-rendering a bunch of things unnecessarily. Is that correct?

Also, when we use it, why are we nesting it inside a `<RenderHTML` rather than a `<Text`?

aldo-expensify replied:
About `TNodeChildrenRenderer`... I actually didn't read the documentation about it, I just saw other usages like here:

https://github.com/Expensify/App/blob/e3e02c99a3c408af6222497e5a1b9327c7839c31/src/components/HTMLEngineProvider/BaseHTMLEngineProvider.js#L69-L111

I just needed something that rendered the rest of the html as it would by default, but allowed me to use the react native styles `mutedTextLabel`. I did the test of removing `'muted-text': TNodeChildrenRenderer,` and it works the same, seems like it is not doing anything useful.

The code 

```
'muted-text': defaultHTMLElementModels.p.extend({
    tagName: 'muted-text',
    mixedUAStyles: styles.mutedTextLabel,
}),
```
I would expect is would make the element `muted-text` render as a `<p>` with our `mutedTextLabel` styles, but in reality this is rendering it as a `<div>`. 

Using `defaultHTMLElementModels.div.extend` broke the layout, that's why I ended up using `p`, which for me made sense as a parent element of a message.

I'll do some testing on an Android device to confirm this is not working only in web.

>Also, when we use it, why are we nesting it inside a <RenderHTML rather than a <Text?

Because the `<Text` doesn't render html

joelbettner replied:
>  I did the test of removing 'muted-text': TNodeChildrenRenderer, and it works the same, seems like it is not doing anything useful.

Yeah...if it is not doing anything then lets get rid of it.

joelbettner reviewed (changes requested):
> I did the test of removing 'muted-text': TNodeChildrenRenderer, and it works the same, seems like it is not doing anything useful.

Yeah...if it is not doing anything then lets get rid of it.

aldo-expensify committed:
Remove useless code - custom renderer

aldo-expensify commented:
Update:

Got rid of `'muted-text': TNodeChildrenRenderer`

Putting this back on WIP for a bit because in Android the link (here) in the error looks miss-aligned 🥲  ... I'll see if there is anything I can do about that!

![Screenshot_20210929-121136](https://user-images.githubusercontent.com/87341702/135334054-a6ada4f7-444f-412e-a0fb-bce22cd08a86.png)

aldo-expensify commented:
The miss-aligned "here" link in the error comes from using a `<PressableWithSecondaryInteraction` to render the anchor tag:

https://github.com/Expensify/App/blob/56374e879bd5d7abedd15017c5c7f35ecc68f531/src/components/AnchorForCommentsOnly/BaseAnchorForCommentsOnly/index.native.js#L39-L60

We end up using this pressable because any anchor `<a>` which `href` is not an internal NewDot link gets rendered as an `AnchorForCommentsOnly`:

https://github.com/Expensify/App/blob/56374e879bd5d7abedd15017c5c7f35ecc68f531/src/components/HTMLEngineProvider/BaseHTMLEngineProvider.js#L92-L110

`BaseHTMLEngineProvider.js` is registering `AnchorRenderer` for `<a>` here:

https://github.com/Expensify/App/blob/56374e879bd5d7abedd15017c5c7f35ecc68f531/src/components/HTMLEngineProvider/BaseHTMLEngineProvider.js#L229-L231

Possible solutions I see:

- This is not an achor in a "comment", so it is probably incorrect to end up rendering this using `AnchorForCommentsOnly`. We could improve our checks [here](https://github.com/Expensify/App/blob/56374e879bd5d7abedd15017c5c7f35ecc68f531/src/components/HTMLEngineProvider/BaseHTMLEngineProvider.js#L81-L90) and make sure we render it using a `Text`
- Fix the rendering of `<AnchorForCommentsOnly` so they don't look miss-aligned when inline with text.
- Do both

aldo-expensify commented:
I see that the miss-aligned link is also happening in the chat for Android:

![Screenshot_20210929-134502](https://user-images.githubusercontent.com/87341702/135346852-e9e05ee5-6a27-4a7f-9bb3-b1af38b3d446.png)

I'll try to do this

> This is not an achor in a "comment", so it is probably incorrect to end up rendering this using AnchorForCommentsOnly. We could improve our checks here and make sure we render it using a Text

And fixing the miss-aligned `<AnchorForCommentsOnly` wont be part of this PR since this is unrelated and was happening before. I'll create a new GH issue for that!

aldo-expensify committed:
Detect rendering anchor within comment

In our HTMLEngineProvider, we are always handling <a> tags using
AnchorRenderer. This renderer will always end up using
AnchorForCommentsOnly if the href of the <a> is not an internal link.

We should not use AnchorForCommentsOnly if this <a> is not within
the context of a comment in a chat.

aldo-expensify commented:
Update: to avoid using `AnchorForCommentsOnly` in contexts that were not comments, we needed someone to recognize that. The cleanest way I found, was this:
- Add new boolean prop `comment` to the `RenderHtml` component. If the `comment` prop is `true`, we wrap the html in a `<div>` with a special attribute `data-comment`. 
- From `AnchorRenderer`, check the ancestor nodes until we find or not find the special attribute. Finding it means that we are rendering inside a comment and we can use `AnchorForCommentsOnly`
- If we don't find it, we default to use a more "default" way of rendering a link:

```
<Text
    style={styles.link}
    onPress={() => Navigation.navigate(internalExpensifyPath)}
>
    <TNodeChildrenRenderer tnode={tnode} />
</Text>
```

aldo-expensify commented:
The miss-alignment link issue was reported here already: https://github.com/Expensify/App/issues/5319

Chatted with @MitchExpensify about it, we will create a new clean GH issue. If this PR goes through, the only miss alignment remaining will be in links sent on the chat.

aldo-expensify committed:
Refactored detection of link in comment

To make it more consistent with <muted-text> tag solution,
I added a <comment> tag to be able to recognize we are rendering
a link within a comment.

RenderHtml doesn't know about comments again.

aldo-expensify commented:
Update:

aldo-expensify commented:
Sorry, closed by mistake...

Update: 

Refactored detection of link in comment

To make it more consistent with <muted-text> tag solution,
I added a <comment> tag to be able to recognize we are rendering 
a link within a comment.

RenderHtml doesn't know about comments again.

Luke9389 commented on src/libs/actions/BankAccounts.js:
```diff
@@ -751,7 +753,9 @@ function setupWithdrawalAccount(data) {
-                    error = response.message;
+                    // Since these specific responses can have an error message in html format with richer content, give priority to the html error.
+                    error = response.htmlMessage || response.message;
+                    htmlError = Boolean(response.htmlMessage);
```
I think you should rename this to something like `isErrorHTML`

aldo-expensify replied:
updated!

Luke9389 commented on src/components/HTMLEngineProvider/BaseHTMLEngineProvider.js:
```diff
@@ -223,6 +258,13 @@ const customHTMLElementModels = {
+        mixedUAStyles: styles.mutedTextLabel,
+    }),
+    comment: defaultHTMLElementModels.div.extend({
+        tagName: 'comment',
```
Do we already use this tag on chat comments?

aldo-expensify replied:
In this PR I'm wrapping the comment's html in:

```
<RenderHTML
    html={`<comment>${fragment.html + (fragment.isEdited ? '<edited></edited>' : '')}</comment>`}
/>
```

Luke9389 commented on src/pages/ReimbursementAccount/ReimbursementAccountForm.js:
```diff
@@ -39,9 +40,15 @@ class ReimbursementAccountForm extends React.Component {
-            );
+            if (this.props.reimbursementAccount.isErrorModalMessageHtml) {
+                error = (
+                    <RenderHTML html={`<muted-text>${this.props.reimbursementAccount.errorModalMessage}</muted-text>`} />
```
So `<muted-text>` is the same as a div with the "muted-text" attribute?

aldo-expensify replied:
Yes, a `<div>` with the styles from `styles.mutedTextLabel` applied.

Luke9389 replied:
Ah ok, so it'll come out to `<div style="styles.mutedTextLabel">`, but `currentNode.domNode.name === muted-text`?

aldo-expensify replied:
> Ah ok, so it'll come out to `<div style="styles.mutedTextLabel">`, but `currentNode.domNode.name === muted-text`?

Yep, but I think you cannot do directly `<div style="styles.mutedTextLabel">`

aldo-expensify committed:
Update variable name

aldo-expensify committed:
Revert unnecessary styling change

aldo-expensify commented on src/components/HTMLEngineProvider/BaseHTMLEngineProvider.js:
```diff
@@ -66,7 +66,28 @@ function computeEmbeddedMaxWidth(tagName, contentWidth) {
+    tnode,
+    key,
+    style,
+}) {
```
updated!

Jag96 commented on src/components/HTMLEngineProvider/BaseHTMLEngineProvider.js:
```diff
@@ -89,6 +106,20 @@ function AnchorRenderer({tnode, key, style}) {
     }
 
+    if (!isInsideComment(tnode)) {
+        // This is not a comment from a chat
```
Can you update this comment to also explain why we want to return something different if the node is not a chat comment?

aldo-expensify replied:
Thanks Joe, I'll check that 😉

aldo-expensify committed:
Merge branch 'main' of github.com:Expensify/App into aldo_vba-add-support-html-error

aldo-expensify committed:
Added comment on using Text for <a>

aldo-expensify committed:
Leave default renderer for tag comment

aldo-expensify commented on src/components/HTMLEngineProvider/BaseHTMLEngineProvider.js:
```diff
@@ -231,6 +269,7 @@ const renderers = {
     code: CodeRenderer,
     img: ImgRenderer,
     edited: withLocalize(EditedRenderer),
+    comment: TNodeChildrenRenderer,
```
`comment: TNodeChildrenRenderer` shouldn't be needed also (removed now). I understand that we don't need any special rendering code for those tags. I think the code

```
    'muted-text': defaultHTMLElementModels.div.extend({
        tagName: 'muted-text',
        mixedUAStyles: styles.mutedTextLabel,
    }),
    comment: defaultHTMLElementModels.div.extend({
        tagName: 'comment',
    }),
```

These tags end up rendering as a `div`, and in the case of `muted-text` we add our `styles.mutedTextLabel` styles.

aldo-expensify committed:
Merge branch 'main' / Resolve conflicts

Luke9389 commented on src/components/HTMLEngineProvider/BaseHTMLEngineProvider.js:
```diff
@@ -66,6 +66,23 @@ function computeEmbeddedMaxWidth(tagName, contentWidth) {
+ * @returns {Boolean}
+ */
+function isInsideComment(tnode) {
+    let currentNode = tnode;
```
What does tnode stand for here? Would you mind renaming this to make that clear?

Luke9389 replied:
Oh, just saw that we have `TNodeRenderer` down there. Nvm

Luke9389 reviewed (approved):
Lookin' good Aldo. Ignore my single comment

Jag96 reviewed (approved):
Tested on web and this LGTM. I noticed that if you click the error message link on the mobile app, you are taken to the sign-in screen on mobile web for OldDot and after signing in the redirect doesn't happen. I don't think that is part of this PR/issue, but just mentioning it in case.
